test/unittests/eof_validation_test.cpp

```diff
@@ -1211,3 +1211,198 @@ TEST_F(eof_validation, max_nested_containers)
+    const auto init_container =
+        eof_bytecode(eofcreate() + returncontract(0, 0, 0), 4).container(runtime_container);
+    const auto factory_code = eofcreate() + OP_STOP;
+    const bytecode factory_container = eof_bytecode(factory_code, 4).container(init_container);
+
+    add_test_case(factory_container, EOFValidationError::incompatible_container_kind);
+}
+
+TEST_F(eof_validation, eofcreate_and_returncontract_return)
+{
+    const auto runtime_container = eof_bytecode(ret(0, 0), 2);
+    const auto init_container =
+        eof_bytecode(eofcreate() + returncontract(0, 0, 0), 4).container(runtime_container);
+    const auto factory_code = eofcreate() + OP_STOP;
+    const bytecode factory_container = eof_bytecode(factory_code, 4).container(init_container);
+
+    add_test_case(factory_container, EOFValidationError::incompatible_container_kind);
+}
+
+TEST_F(eof_validation, eofcreate_and_returncontract_returncontract)
+{
+    const auto runtime_container =
+        eof_bytecode(returncontract(0, 0, 0), 2).container(eof_bytecode(OP_INVALID));
+    const auto init_container =
+        eof_bytecode(eofcreate() + returncontract(0, 0, 0), 4).container(runtime_container);
+    const auto factory_code = eofcreate() + OP_STOP;
+    const bytecode factory_container = eof_bytecode(factory_code, 4).container(init_container);
+
+    add_test_case(factory_container, EOFValidationError::incompatible_container_kind);
+}
+
+TEST_F(eof_validation, eofcreate_and_returncontract_revert)
+{
+    const auto runtime_container = eof_bytecode(revert(0, 0), 2);
+    const auto init_container =
+        eof_bytecode(eofcreate() + returncontract(0, 0, 0), 4).container(runtime_container);
+    const auto factory_code = eofcreate() + OP_STOP;
+    const bytecode factory_container = eof_bytecode(factory_code, 4).container(init_container);
+
+    add_test_case(factory_container, EOFValidationError::success);
+}
+
+TEST_F(eof_validation, eofcreate_stop_and_returncontract)
+{
+    const auto runtime_container = eof_bytecode(OP_INVALID);
+    const auto init_code = rjumpi(1, 0) + OP_STOP + returncontract(0, 0, 0);
+    const auto init_container = eof_bytecode(init_code, 2).container(runtime_container);
+    const auto factory_code = eofcreate() + OP_STOP;
+    const bytecode factory_container = eof_bytecode(factory_code, 4).container(init_container);
+
+    add_test_case(factory_container, EOFValidationError::ambiguous_container_kind);
+}
+
+TEST_F(eof_validation, eofcreate_return_and_returncontract)
+{
+    const auto runtime_container = eof_bytecode(OP_INVALID);
+    const auto init_code = rjumpi(5, 0) + ret(0, 0) + returncontract(0, 0, 0);
+    const auto init_container = eof_bytecode(init_code, 2).container(runtime_container);
+    const auto factory_code = eofcreate() + OP_STOP;
+    const bytecode factory_container = eof_bytecode(factory_code, 4).container(init_container);
+
+    add_test_case(factory_container, EOFValidationError::ambiguous_container_kind);
+}
```

Comment:
I would also add a test where the offending terminating opcode is found in a non-first code section.

Reply:
I also had in mind to add some tests with invalid combinations on a deeper nesting level.